syntax = "proto3";

package user;

import "google/protobuf/timestamp.proto";

// 1. THE ENTITY (Mirrors the database table)
message User {
  string id = 1;
  string email = 2;

  // Name fields
  string first_name = 3;
  optional string middle_name = 4;
  string last_name = 5;

  // Security fields
  string password_hash = 6;
  AuthProvider auth_provider = 7;
  optional string auth_subject = 8;

  // State fields
  UserStatus status = 9;
  bool mfa_enabled = 10;

  // Preference fields
  optional string timezone = 11;
  optional string locale = 12;
  optional string avatar_url = 13;

  // Default is "" -> SQL default is {}
  string settings_json = 14;


  // Timestamps
  optional google.protobuf.Timestamp email_verified_at = 15;
  optional google.protobuf.Timestamp last_login_at = 16;
  google.protobuf.Timestamp created_at = 17;
  google.protobuf.Timestamp updated_at = 18;
  optional google.protobuf.Timestamp deleted_at = 19;
}

// 2. ENUMS
enum UserStatus {
  STATUS_UNSPECIFIED = 0;
  ACTIVE = 1;
  INACTIVE = 2;
  SUSPENDED = 3;
  DELETED = 4;
}

enum AuthProvider {
  LOCAL = 0;
  GOOGLE = 1;
  MICROSOFT = 2;
  APPLE = 3;
}

// SERVICE & DTO's
service UserService {
  rpc GetUserById(GetUserByIdRequest) returns (UserResponse);
  rpc GetUserByEmail(GetUserByEmailRequest) returns (UserResponse);
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
}

message CreateUserResponse {
  bool success = 1;
}

message GetUserByIdRequest {
  string id = 1;
}

message GetUserByEmailRequest {
  string email = 1;
}

message UserResponse {
  User user = 1;
}

message CreateUserRequest {
  string email = 1;
  string first_name = 2;
  optional string middle_name = 3;
  string last_name = 4;
  string password = 5;
  AuthProvider auth_provider = 6;
  optional string auth_subject = 7;
  optional string timezone = 8;
  optional string locale = 9;
  optional string avatar_url = 10;
}